Lighting/Src/lighting_controller.cpp

```diff
@@ -308,41 +504,118 @@ void HAL_TIM_PWM_PulseFinishedCallback(TIM_HandleTypeDef *htim) {
+	if (rev4.get_drone_state() == GROUND_STATE && g_num_seconds == 5) {
+		if (enter_standby == false) {
+			rev4.transition_to(2);	//go to TAXI_STATE
+			rev4.activate_domain(CD_TAXI);
+			enter_standby = true;
+		} else {
+			rev4.transition_to(1);	//stay in GROUND_STATE but show standby lighting pattern.
+			enter_standby = false;
+		}
+		g_num_seconds = 0;
+	} else if (rev4.get_drone_state() == TAXI_STATE && g_num_seconds == 1) {
+		rev4.transition_to(1);	//go to TAKE_OFF_STATE
+		g_num_seconds = 0;
+	} else if(rev4.get_drone_state() == TAKE_OFF_STATE && g_num_seconds == 1) {
+		rev4.transition_to(1);	//go to FLIGHT_STATE
+		rev4.activate_domain(CD_NAVIGATION);
+		g_num_seconds = 0;
+	} else if (rev4.get_drone_state() == FLIGHT_STATE && g_num_seconds == 5) {
+		rev4.transition_to(1);	//go to LANDING_STATE
+		rev4.activate_domain(CD_LANDING);
+		g_num_seconds = 0;
+	} else if (rev4.get_drone_state() == LANDING_STATE && g_num_seconds == 5) {
+		rev4.transition_to(0);	//go to SEARCH_STATE
+		rev4.activate_domain(CD_BEACON);
+		g_num_seconds = 0;
+	} else if (rev4.get_drone_state() == SEARCH_STATE && g_num_seconds == 5) {
+		rev4.transition_to(1);	//go to GROUND_STATE (no standby)
+		rev4.activate_domain(CD_BEACON);
+		g_num_seconds = 0;
+	}
+
+	g_num_seconds += 1;
 }
 
 void TIM7_100msCallback(TIM_HandleTypeDef *htim7) {
 	static uint8_t stage = 0;
+	static uint8_t led_index = 0;
+
 	if (stage == 0) { 			// STROBE ON
-		rev3.enable_control_domain(CD_STROBE);
+		rev4.activate_domain(CD_STROBE);
 		stage = 1;
 	} else if (stage == 1) { 	// STROBE OFF
-		rev3.disable_control_domain(CD_STROBE);
+		rev4.deactivate_domain(CD_STROBE);
 		stage = 2;
 	} else if (stage == 2) {	// STROBE ON
-		rev3.enable_control_domain(CD_STROBE);
+		rev4.activate_domain(CD_STROBE);
 		stage = 3;
 	} else if (stage == 3) {	// STROBE OFF
-		rev3.disable_control_domain(CD_STROBE);
+		rev4.deactivate_domain(CD_STROBE);
 		stage = 4;
 	} else if (stage == 4) {	// OFF
 		stage = 5;
 	} else if (stage == 5) {	// OFF
 		stage = 6;
 	} else if (stage == 6) {	// BCN ON
-		rev3.enable_control_domain(CD_BEACON);
+		if (rev4.get_drone_state() >= FLIGHT_STATE) {
+			rev4.activate_domain(CD_BEACON);
+		}
 		stage = 7;
 	} else if (stage == 7) {	// BCN OFF
-		rev3.disable_control_domain(CD_BEACON);
+		if (rev4.get_drone_state() >= FLIGHT_STATE) {
+			rev4.deactivate_domain(CD_BEACON);
+		}
 		stage = 0;
 		HAL_TIM_Base_Stop_IT(htim7);
 	}
+
+	//"Scrolling" pattern for the "Search" drone state.
+	if (rev4.get_drone_state() == SEARCH_STATE) {
+		uint8_t num_top_facing_leds = 6;
+		if (led_index == 0) {
+			rev4.remove_led_from_cd(num_top_facing_leds-1, CD_SEARCH);
+		} else {
+			rev4.remove_led_from_cd(led_index-1, CD_SEARCH);
+		}
+
+		rev4.add_led_to_cd(led_index, CD_SEARCH);
+		rev4.activate_domain(CD_SEARCH);
+		led_index = (led_index + 1) % num_top_facing_leds;
+	}
 }
 
-void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
+void TIM2_20msCallback(TIM_HandleTypeDef *htim2) {
```

Comment:
this gets a pass for now since we aren't using the timer for anything else, but it is almost always better to have functions like this exist inside a `main()` loop, and not inside an ISR.

there's not really a good argument for having your colour cycling startup sequence so so tightly controlled by the timer, so it's totally possible to get away with using `HAL_Delay` or `wait_until()` statements in your main loop. All you need to do is sit domain brightnesses for CD_MAIN.